scalyr_agent/util.py

```diff
@@ -1337,3 +1341,265 @@ def close(self):
+                                self._current_cluster_rate < self._initial_cluster_rate):
+                            self._current_cluster_rate = self._initial_cluster_rate
+                        else:
+                            self._current_cluster_rate = min(
+                                self._upper_cluster_rate, self._current_cluster_rate * self._increase_factor)
+                    self._consecutive_successes = 0
+        finally:
+            self._cluster_rate_lock.release()
+
+    def _get_next_ripe_time(self):
+        """Get the absolute time for when this token becomes next available
+
+        A token contains state representing the last time it was acquired.
+        In order to achieve a uniform cluster-wide average rate of R, the next time is calculated as:
+
+        current_time + delta
+
+        where delta = random number between (0, 2 * T)
+        where T is the per-agent interval or num_agents / cluster rate
+        """
+        agent_interval = float(self._num_agents) / self._current_cluster_rate
+        delta = random.uniform(0, 2 * agent_interval)
+        t = self._time() + delta
+        return t
+
+    def _time(self):
+        """Returns absolute time in UTC epoch seconds"""
+        if self._fake_clock:
+            return self._fake_clock.time()
+        else:
+            return time.time()
+
+    def _simulate_acquire_token(self):
+        self._test_mode_lock.acquire()
+        try:
+            while len(self._token_queue) == 0 or self._time() < self._ripe_time:
+                self._test_mode_lock.release()
+                self._fake_clock.simulate_waiting()
+                self._test_mode_lock.acquire()
+
+            # Head token is ripe
+            token = self._token_queue.popleft()
+            self._ripe_time = self._get_next_ripe_time()
+            return token
+        finally:
+            self._test_mode_lock.release()
+
+    def _simulate_release_token(self, token, success):
+        self._adjust_cluster_rate(success)
+        self._test_mode_lock.acquire()
+        try:
+            self._token_queue.append(token)
+        finally:
+            self._test_mode_lock.release()
+        self._fake_clock.advance_time(increment_by=self._ripe_time - self._fake_clock.time())
+
+    def acquire_token(self):
+        """Acquires a token, blocking until a token becomes available.
+
+        The primary state used to determine token availability is self._ripe_time. Until that time is reached,
+        all client threads are blocked.
+
+        @return: a token object
+        @rtype: RateLimiterToken
+        """
+        if self._fake_clock:
+            return self._simulate_acquire_token()
+
+        # block until a token is available from the token heap
+        self._token_queue_cv.acquire()
+
+        try:
+            while len(self._token_queue) == 0 or self._time() < self._ripe_time:
```

Comment:
I think it would be better if you broke this up into two steps.. first, get a token from the token queue... you have to be holding the lock while you do that.  However, after you get the token (and record the ripe time when you get that token).. then you sleep until the ripe time has past.  You can do that without holding onto the token_queue_cv lock.

I think there are a couple of advantages with this approach.  First, in my mind simplifies the code here because you aren't trying to do two things in this loop.  Second, you don't have to call `_time()` while holding the lock... getting the system time can be weird now and then.. it is a system call and it does more than you'd think.  I try to avoid holding a lock if I have to do that.

I think the other change you'd have to make if you went this route is change `_get_next_ripe_time`.  Currently, you based the next ripe time on the current time, which works if you only invoke `_get_next_ripe_time` when the last ripe time has been reached.  With my approach, you'd be calling it when you acquire the token but before you actually sleep (to wait until the current time is ripe time).  But, I think that is easy to fix as well -- have `_get_next_ripe_time` be based on the max(current_time, last_ripe_time.)  you want to still use current time to cover the case when a thread calls `acquire_token` with the ripe time already past.

Reply:
@czerwingithub I'd like to understand if you see this change as a must-have.  I've spent too many hours trying to implement what I think satisfies your recommended change here: https://github.com/scalyr/scalyr-agent-2/pull/229

The problem I'm having is getting my tests to pass for multithreaded situation. I know I'm getting race conditions. What I don't know if it's just because of the way my test is written (to use the fake clock) of whether these will manifest in production code.  

Due to time pressure, unless you see this change as necessary, I would rather punt on this as there are other pressing things to work on. (iiuc your main concern is the extra cycles it takes to call system clock, but would that amount to *significant* problems in any scenario?)